File: src/main/scala/ProbFunctions.scala
import scala.math._
/**
 * Utility functions for calculating implied probabilities from different odds formats.
 */
object ProbFunctions {

  /**
   * Rounds a floating-point number to two decimal places.
   *
   * @param floatNum The input floating-point number.
   * @return The rounded floating-point number.
   */
  private def roundToSecondDecimal(floatNum: Float): Float = (floatNum * 100).round.toFloat / 100.0f

  /**
   * Calculate the implied probability from Moneyline odds.
   *
   * @param moneyline The Moneyline odds as a Float.
   * @return The implied probability as a Float, rounded to two decimal places.
   */
  def moneylineProb(moneyline: Float): Float = moneyline match {
    case moneylineValue if moneylineValue > 0 => roundToSecondDecimal((100 / (moneyline + 100.0f)) * 100)
    case _ => roundToSecondDecimal(((moneyline * -1) / ((moneyline * -1) + 100)) * 100)
  }

  /**
   * Calculate the implied probability from Decimal odds.
   *
   * @param decimal The Decimal odds as a Float.
   * @return The implied probability as a Float, rounded to two decimal places.
   */
  def decimalProb(decimal: Float): Float = roundToSecondDecimal((1.0f / decimal)* 100)

  /**
   * Calculate the implied probability from Fractional odds.
   *
   * @param fractions The Fractional odds as a Float.
   * @return The implied probability as a Float, rounded to two decimal places.
   */
  def fractionProb(fractions: Float): Float = roundToSecondDecimal((1.0f / fractions)* 100)

}

File: src/prob_functions.rs
fn round_to_two_decimal_places(float_number :f32) -> f32{
    (float_number * 100.0).round() / 100.0
}

pub fn moneyline_prob(moneyline :f32) -> f32 {
    let probability: f32= match moneyline {
        moneyline_value if moneyline_value > 0.0 => (100.0 / (moneyline + 100.0)) * 100.0,
        _ => ((moneyline.abs()) / (moneyline.abs() + 100.0)) * 100.0
    };

    return round_to_two_decimal_places(probability);
}

fn calculate_prob(odds: f32) -> f32 {
    round_to_two_decimal_places((1.0 / odds) * 100.0)
}

pub fn decimal_prob(decimal: f32) -> f32 {
    calculate_prob(decimal)
}

pub fn fraction_prob(fractions: f32) -> f32 {
    calculate_prob(fractions)
}

pub fn calculate_payout(implied_prob : f32, wager: f64) -> f64 {
 let odds : f64 =  1.0 / (implied_prob as f64);

 odds * wager
}


#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_round_to_two_decimal_places() {
        assert_eq!(round_to_two_decimal_places(1.23456), 1.23);
        assert_eq!(round_to_two_decimal_places(1.235), 1.24);
        assert_eq!(round_to_two_decimal_places(100.0), 100.0);
    }

    #[test]
    fn test_moneyline_prob_positive() {
        let moneyline: f32 = 150.0;
        let expected_prob: f32 = (100.0 / (moneyline + 100.0)) * 100.0;
        assert_eq!(moneyline_prob(moneyline), round_to_two_decimal_places(expected_prob));
    }

    #[test]
    fn test_moneyline_prob_negative() {
        let moneyline: f32 = -150.0;
        let expected_prob: f32 = (moneyline.abs() / (moneyline.abs() + 100.0)) * 100.0;
        assert_eq!(moneyline_prob(moneyline), round_to_two_decimal_places(expected_prob));
    }

    #[test]
    fn test_decimal_prob() {
        let decimal: f32 = 2.5;
        let expected_prob: f32 = (1.0 / decimal) * 100.0;
        assert_eq!(decimal_prob(decimal), round_to_two_decimal_places(expected_prob));
    }

    #[test]
    fn test_fraction_prob() {
        let fractions: f32 = 5.0 / 2.0;
        let expected_prob: f32 = (1.0 / fractions) * 100.0;
        assert_eq!(fraction_prob(fractions), round_to_two_decimal_places(expected_prob));
    }
}